runtime/lua/vim/treesitter/query.lua

```diff
@@ -175,11 +173,22 @@ function M.get_node_text(node, source)
   local end_row, end_col, end_byte = node:end_()
 
   if type(source) == "number" then
-    if start_row ~= end_row then
-      return nil
+    local lines
+    if end_col == 0 then
+      lines = a.nvim_buf_get_lines(source, start_row, end_row, true)
```

Comment:
You'll need to check for start/end row not being larger than the number of lines of the buffer, similar to https://github.com/neovim/neovim/pull/15030/

Reply:
Do you have an example where this happens? I'd like to create a test case for it

Reply:
he, this is kind of hard to replicate with the c parser, as it needs to have a node that includes the EOF mark at the start and/or end of de node. I was using this to replicate cases like this


```lua
local parsers = require "nvim-treesitter.parsers"
local query = require "vim.treesitter.query"

local function print(str)
  vim.api.nvim_err_writeln(vim.inspect(str))
end

local function show(node)
  print { node:start() }
  print { node:end_() }
  print(query.get_node_text(node, 0))
end

local lang = "ruby"

local parser = parsers.get_parser(0, lang)
local tree = parser:parse()[1]
local node = tree:root()
local q = query.parse_query(lang, "(heredoc_end) @this")
for pattern, match, metadata in q:iter_matches(node, 0, 1, 4) do
  for id, n in pairs(match) do
    show(n)
  end
end
```

Executed while having this file open

```ruby
def run
  a = <<~E
end
```

I think is easier to test this mocking the node, but don't know if that's possible to do in those tests.

Reply:
Thanks. I've added a test case